### PR TITLE
etcd boostrapping using dns

### PR DESCRIPTION
**Pull request by abhinavdahiya**
1. cmd: add setup-etcd-environment

This outputs an environment file with envs that allow etcd to discover its ip address
and the corresponding dns name from the discovery-srv records.

It looks up the `_etcd-server-ssl._tcp.<domain specified by --discovery-srv>` for reverse lookup
of its own dns name.

An example of such file is
```console
$ cat /etc/etcd.env/etcd-environment
ETCD_IPV4_ADDRESS=192.168.126.11
ETCD_DNS_NAME=adahiya-0-etcd-0.tt.testing
```
when the `--discovery-srv` was `tt.testing`
```console
dig +noall +answer _etcd-server-ssl._tcp.tt.testing SRV
_etcd-server-ssl._tcp.tt.testing. 0 IN  SRV     0 10 2380 adahiya-0-etcd-0.tt.testing.
```

The command retries reverse lookup until 5 minutes to allow for dns to be available.

2.  templates: bootstrap etcd using dns discovery

Using the guide https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/clustering.md#dns-discovery

Also adds `setup-etcd-environment.service` that uses the setup-etcd-environment cli to setup the discovery params.

3. cleanup across controller, server and operator to drop etcd index, etcd initial count.

This requires https://github.com/openshift/installer/pull/526

### CODE REVIEW

**Comment by abhinavdahiya** [comment]
/cc @crawford @aaronlevy 

This seem reasonable?

**Comment by ashcrow** [comment]
FWIW @sdemos opened up a request as to why e2e-aws hangs for 2+ hours and then fails. We're seeing it on other PRs as well.

File: templates/_base/master/units/etcd-member.yaml [hunk 1]
```diff
@@ -6,19 +6,22 @@ contents: |
...
   [Service]
   Restart=on-failure
   RestartSec=10s
   TimeoutStartSec=0
   LimitNOFILE=40000
 
+  ## FIXME(abhinav): these images should be replacable by release image.
   Environment="SIGNER_IMAGE=quay.io/coreos/kube-client-agent:678cc8e6841e2121ebfdb6e2db568fce290b67d6"
   Environment="ETCD_IMAGE=quay.io/coreos/etcd:v3.2.14"
+  EnvironmentFile=/etc/etcd.env/etcd-environment
```
Comment (crawford) [inline templates/_base/master/units/etcd-member.yaml]: `/run/etcd/environment`

File: templates/_base/master/units/setup-etcd-environment.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+  ExecStartPre=/usr/bin/mkdir --parents /etc/etcd.env
+  ExecStart=/bin/podman \
+    run \
+      --net host \
+      --rm \
+      --volume /etc/etcd.env:/etc/etcd.env:z \
+      ${IMAGE} \
+        --discovery-srv={{.BaseDomain}} \
+        --output-file=/etc/etcd.env/etcd-environment \
```
Comment (crawford) [inline templates/_base/master/units/setup-etcd-environment.yaml]: `/run/etcd/environment`

File: pkg/controller/template/render.go [hunk 3]
```diff
@@ -294,53 +293,31 @@ func skipMissing(key string) (interface{}, error) {
...
 
 	var dnsNames = []string{
-		"*.kube-etcd.kube-system.svc.cluster.local",
-		"kube-etcd-client.kube-system.svc.cluster.local",
-	}
-
-	for i := 0; i < cfg.EtcdInitialCount; i++ {
-		dnsNames = append(dnsNames, fmt.Sprintf("%s-etcd-%d.%s", cfg.ClusterName, i, cfg.BaseDomain))
+		"${ETCD_DNS_NAME}",
+		cfg.BaseDomain, // https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/clustering.md#dns-discovery
```
Comment (crawford) [inline pkg/controller/template/render.go]: Instead of master, use a specific commit so that document doesn't disappear on us.

**Comment by crawford** [comment]
/lgtm
/hold

Waiting on https://github.com/openshift/installer/pull/526.

**Comment by crawford** [comment]
/hold cancel

CI will prevent us from merging this too early.

**Comment by abhinavdahiya** [comment]
[e2e-aws](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/143/pull-ci-openshift-machine-config-operator-master-e2e-aws/153) tests completed, teardown hung :(
```console
2018/10/27 07:48:54 Container setup in pod e2e-aws completed successfully
2018/10/27 08:01:07 Container test in pod e2e-aws completed successfully
2018/10/27 10:41:04 Copying artifacts from e2e-aws into /logs/artifacts/e2e-aws
2018/10/27 10:41:05 error: unable to signal to artifacts container to terminate in pod e2e-aws, triggering deletion: could not run remote command: unable to upgrade connection: container not found ("artifacts")
2018/10/27 10:41:05 error: unable to retrieve artifacts from pod e2e-aws: could not read gzipped artifacts: unable to upgrade connection: container not found ("artifacts")
E1027 10:41:09.973225      11 event.go:200] Server rejected event '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:".1561711212047510", GenerateName:"", Namespace:"ci-op-zp6jy1rz", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, InvolvedObject:v1.ObjectReference{Kind:"", Namespace:"ci-op-zp6jy1rz", Name:"", UID:"", APIVersion:"", ResourceVersion:"", FieldPath:""}, Reason:"CiJobFailed", Message:"Running job pull-ci-openshift-machine-config-operator-master-e2e-aws for PR https://github.com/openshift/machine-config-operator/pull/143 in namespace ci-op-zp6jy1rz from author abhinavdahiya", Source:v1.EventSource{Component:"ci-op-zp6jy1rz", Host:""}, FirstTimestamp:v1.Time{Time:time.Time{wall:0xbeed2e1179670310, ext:11024912575830, loc:(*time.Location)(0x1973400)}}, LastTimestamp:v1.Time{Time:time.Time{wall:0xbeed2e1179670310, ext:11024912575830, loc:(*time.Location)(0x1973400)}}, Count:1, Type:"Warning", EventTime:v1.MicroTime{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, Series:(*v1.EventSeries)(nil), Action:"", Related:(*v1.ObjectReference)(nil), ReportingController:"", ReportingInstance:""}': 'events ".1561711212047510" is forbidden: unable to create new content in namespace ci-op-zp6jy1rz because it is being terminated' (will not retry!)
```
/retest

**Comment by abhinavdahiya** [comment]
```console
level=fatal msg="Error executing openshift-install: RequestLimitExceeded: Request limit exceeded.\n\tstatus code: 503, request id: 363ae6fb-81bc-40c2-b837-cc5b03692f4e"
2018/10/27 19:21:20 Container setup in pod e2e-aws failed, exit code 1, reason Error
Another process exited
```
Whats up with our ci? :(

/retest

**Comment by ashcrow** [comment]
Yeah something is up. We've been trying to merge a PR for a bit over a week :-(.

**Pull request merged**